### PR TITLE
[WIP] Switch exercise projects to self-hosting NServiceBus

### LINKED ISSUE #115: Move off NServiceBus.Host
given it will shortly be deprecated

### PR DESCRIPTION
**Pull request by seanfarmar**
Fixes #115

### CODE REVIEW

**Comment by seanfarmar** [comment]
@mauroservienti @adamralph ping

**Comment by seanfarmar** [comment]
Second version - converted the project to self-hosting.

@Particular/microservices-workshop-maintainers - Please review and if it's ok I'll do all the rest

File: exercises/04-integration/after/04-integration-after.sln [hunk 1]
```diff
@@ -1,10 +1,7 @@
-﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 14
-VisualStudioVersion = 14.0.25420.1
+# Visual Studio 15
+VisualStudioVersion = 15.0.26730.16
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Divergent.Customers", "Divergent.Customers\Divergent.Customers.csproj", "{02EE3114-734F-42DF-994C-98480E1489FE}"
-EndProject
```
Comment (adamralph) [inline exercises/04-integration/after/04-integration-after.sln]: It looks like this file has become polluted due to the addition of `Customers.Host` in the first commit, and then the removal of the previous `Customers` and rename of `Customers.Host` to `Customers` in the second commit.

I think it's better if we treat this branch as a spike, to get feedback, but then the actual work is started again, on a fresh branch and in a new PR.

File: exercises/04-integration/after/Divergent.Customers/Config/EndpointConfig.cs [hunk 2]
```diff
@@ -48,8 +32,12 @@ public void Customize(EndpointConfiguration endpointConfiguration)
...
 
             var conventions = endpointConfiguration.Conventions();
-            conventions.DefiningCommandsAs(t => t.Namespace != null && t.Namespace.StartsWith("Divergent") && t.Namespace.EndsWith("Commands") && t.Name.EndsWith("Command"));
-            conventions.DefiningEventsAs(t => t.Namespace != null && t.Namespace.StartsWith("Divergent") && t.Namespace.EndsWith("Events") && t.Name.EndsWith("Event"));
+            conventions.DefiningCommandsAs(t =>
+                t.Namespace != null && t.Namespace.StartsWith("Divergent") && t.Namespace.EndsWith("Commands") &&
+                t.Name.EndsWith("Command"));
+            conventions.DefiningEventsAs(t =>
+                t.Namespace != null && t.Namespace.StartsWith("Divergent") && t.Namespace.EndsWith("Events") &&
+                t.Name.EndsWith("Event"));
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Customers/Config/EndpointConfig.cs]: It took me a while to figure out that there are no actual changes here and it's just the addition of line breaks. This is a good change, but please make it in a separate refactoring commit if it is going to be in the new PR.

File: exercises/04-integration/after/Divergent.Customers/ProgramService.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        static ProgramService()
+        {
+            //TODO: optionally choose a custom logging library
+            // https://docs.particular.net/nservicebus/logging/#custom-logging
+            // LogManager.Use<TheLoggingFactory>();
+            logger = LogManager.GetLogger<ProgramService>();
+        }
+
+        public static void Main()
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: I'd prefer not to conflate `Program.Main` and the `ServiceBase`. Can we keep them separate?

File: exercises/04-integration/after/Divergent.Customers/ProgramService.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+using System.ServiceProcess;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Logging;
+using Divergent.Customers.Config;
+
+namespace Divergent.Customers
+{
+    [DesignerCategory("Code")]
+    class ProgramService : ServiceBase
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: `ProgramService` is a weird name. I'd prefer `WindowsService`.

File: exercises/04-integration/after/Divergent.Customers/ProgramService.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+using System.ComponentModel;
+using System.ServiceProcess;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Logging;
+using Divergent.Customers.Config;
+
+namespace Divergent.Customers
+{
+    [DesignerCategory("Code")]
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: What is this?

File: exercises/04-integration/after/Divergent.Customers/ProgramService.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.ComponentModel;
+using System.ServiceProcess;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Logging;
+using Divergent.Customers.Config;
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: Needs sorting.

File: exercises/04-integration/after/Divergent.Customers/ProgramService.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        IEndpointInstance endpoint;
+        static ILog logger;
+
+        static ProgramService()
+        {
+            //TODO: optionally choose a custom logging library
+            // https://docs.particular.net/nservicebus/logging/#custom-logging
+            // LogManager.Use<TheLoggingFactory>();
+            logger = LogManager.GetLogger<ProgramService>();
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: This should be a field initialiser rather than in a static constructor.

File: exercises/04-integration/after/Divergent.Customers/ServiceHelperClassic.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+namespace Divergent.Customers
+{
+    static class ServiceHelper
+    {
+        public static bool IsService()
+        {
+            return !Environment.UserInteractive;
+        }
+    }
+}
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Customers/ServiceHelperClassic.cs]: This is used in one place only. I'd prefer to get rid of it and have the caller call `Environment.UserInteractive` instead.

File: exercises/04-integration/after/Divergent.Customers/ProgramService.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+using NServiceBus.Logging;
+using Divergent.Customers.Config;
+
+namespace Divergent.Customers
+{
+    [DesignerCategory("Code")]
+    class ProgramService : ServiceBase
+    {
+        IEndpointInstance endpoint;
+        static ILog logger;
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: Both these should be prefixed by `private`.

Comment (mauroservienti) [reply exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: I'm used to prefix with private as well, so I'm fine with the suggestion. But isn't it redundant? fields are private by default.

**Comment by mauroservienti** [comment]
> In addition to the inline comments, it looks ProgramService is based on some template with a lot of redundant comments and (sometimes empty) private methods. That should all be cleaned up.

I suppose that the template comes from https://github.com/Particular/NServiceBus.Bootstrap.WindowsService

File: exercises/04-integration/after/Divergent.Customers/ProgramService.cs (hunk 9 above)
Comment (adamralph) [reply exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: I prefer to always be explicit. Different types of members have different access defaults and I'm loathed to have to map back to the default in my mind while I'm reading the code.

Comment (mauroservienti) [reply exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: `class` should be `internal` then 😉

Comment (adamralph) [reply exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: Yep, good catch! 😄

File: exercises/04-integration/after/04-integration-after.sln (hunk 1 above)
Comment (seanfarmar) [reply exercises/04-integration/after/04-integration-after.sln]: I will reset the head and do a clean conversion 👍

File: exercises/04-integration/after/Divergent.Customers/ProgramService.cs (hunk 5 above)
Comment (seanfarmar) [reply exercises/04-integration/after/Divergent.Customers/ProgramService.cs]: It comes from the template, I can clean all this if we need to

File: exercises/04-integration/after/Divergent.Customers/ServiceHelperClassic.cs (hunk 8 above)
Comment (seanfarmar) [reply exercises/04-integration/after/Divergent.Customers/ServiceHelperClassic.cs]: @bording did a [PR to clean this template](https://github.com/Particular/dotnetTemplates/pull/3)

**Comment by bording** [comment]
@adamralph as @seanfarmar mentioned, most of the things you're mentioning here come from the `dotnet new` template here: https://github.com/Particular/dotnetTemplates/blob/develop/src/Templates/NServiceBusWindowsService/ProgramService.cs

I've already made some changes, but if you think there are other things that we should alter as part of the template we'll be providing to customers, please suggest changes there.

File: exercises/02-publish-subscribe/after/Divergent.Customers/Handlers/OrderSubmittedHandler.cs [hunk 10]
```diff
@@ -1,9 +1,9 @@
 ﻿using System.Data.Entity;
 using System.Threading.Tasks;
-using Divergent.Customers.Data.Context;
-using Divergent.Sales.Messages.Events;
 using NServiceBus;
 using NServiceBus.Logging;
+using Divergent.Customers.Data.Context;
+using Divergent.Sales.Messages.Events;
```
Comment (adamralph) [inline exercises/02-publish-subscribe/after/Divergent.Customers/Handlers/OrderSubmittedHandler.cs]: This is an unrelated change. It should really be in a separate commit or PR. Same in a few other places.

File: exercises/02-publish-subscribe/after/Divergent.Customers/Config/EndpointConfig.cs [hunk 11]
```diff
@@ -1,32 +1,25 @@
...
 using NServiceBus.Logging;
 using NServiceBus.Persistence;
-using ILog = Common.Logging.ILog;
-using LogManager = Common.Logging.LogManager;
 
 namespace Divergent.Customers.Config
 {
-    [EndpointName("Divergent.Customers")]
-    public class EndpointConfig : IConfigureThisEndpoint, AsA_Server
+    public class EndpointConfig
```
Comment (adamralph) [inline exercises/02-publish-subscribe/after/Divergent.Customers/Config/EndpointConfig.cs]: This class should be static (same in other projects).

**Comment by dvdstelt** [comment]
I definitely think we should definitely keep supporting VS2015 for as long as we can. Not everyone has the opportunity to upgrade because of various reasons.

**Comment by adamralph** [comment]
At BuildStuff we still had some people (although a small minority) using VS 2015. I believe we should probably keep supporting it for now. Not everyone has a personal laptop where they can install VS 2017, nor a work laptop where VS 2017 is permitted (some orgs are still very slow to adopt new IDE versions).

**Comment by mauroservienti** [comment]
Agreed. What's the impact on .NET Core projects, if any?

> Visual Studio 2015 will only support .NET Core <= 1.1 (ie. project.json). If you wish to use .csproj based .NET Core projects you will need to use Visual Studio 2017

**Comment by bording** [comment]
> Agreed. What's the impact on .NET Core projects, if any?

@mauroservienti You have to have VS 2017 for .NET Core projects. AFAIK the old VS 2015 tooling cannot be used for a .NET Core 2.0 project.

**Comment by adamralph** [comment]
As [discussed](https://github.com/Particular/Workshop/pull/193#issuecomment-350239333), we'll hold off merging this until after the merge of https://github.com/Particular/Workshop/pull/193

**Comment by adamralph** [comment]
Superseded by #210
